cli/compiler.rs

```diff
@@ -322,53 +324,74 @@ pub fn compile_sync(
 
-  match union.wait() {
-    Ok((result, i, rest)) => {
-      // We got a sucessful finish before any recivers where canceled
-      let mut rest_mut = rest;
-      match ((*result.deref()).clone(), i) {
-        // Either receiver was completed with success.
-        (Ok(v), _) => Ok(v),
-        // Either receiver was completed with a valid error
-        // this should be fatal for now since it is not intended
-        // to be possible to recover from a uncaught error in a isolate
-        (Err(Some(err)), _) => Err(err),
-        // local_receiver finished first with a none error. This is intended
-        // to catch when the local logic can't complete because it is unable
-        // to send and/or receive messages from the compiler worker.
-        // Due to the way that scheduling works it is very likely that the
-        // compiler worker future has already or will in the near future
-        // complete with a valid JSError or a None.
-        (Err(None), 1) => {
-          debug!("Compiler local exited with None error!");
-          // While technically possible to get stuck here indefinately
-          // in theory it is highly unlikely.
-          debug!(
-            "Waiting on compiler worker result specifier: {} referrer: {}!",
-            specifier, referrer
-          );
-          let worker_result =
-            (*rest_mut.remove(0).wait().unwrap().deref()).clone();
-          debug!(
-            "Finished waiting on worker result specifier: {} referrer: {}!",
-            specifier, referrer
-          );
-          match worker_result {
-            Err(Some(err)) => Err(err),
-            Err(None) => panic!("Compiler exit for an unknown reason!"),
-            Ok(v) => Ok(v),
+  let specifier_ = specifier.to_string().clone();
+  let referrer_ = specifier.to_string().clone();
+
+  union.then(move |result| {
+    match result {
+      Ok((result, i, rest)) => {
+        // We got a sucessful finish before any recivers where canceled
+        let mut rest_mut = rest;
+        match ((*result.deref()).clone(), i) {
+          // Either receiver was completed with success.
+          (Ok(v), _) => Either::A(futures::future::result(Ok(v))),
+          // Either receiver was completed with a valid error
+          // this should be fatal for now since it is not intended
+          // to be possible to recover from a uncaught error in a isolate
+          (Err(Some(err)), _) => Either::A(futures::future::result(Err(err))),
+          // local_receiver finished first with a none error. This is intended
+          // to catch when the local logic can't complete because it is unable
+          // to send and/or receive messages from the compiler worker.
+          // Due to the way that scheduling works it is very likely that the
+          // compiler worker future has already or will in the near future
+          // complete with a valid JSError or a None.
```

Comment:
Can this be demonstrated with a unit test below?

I’m all for this PR and will land it but I need to spend some time with this code to wrap my head around it. It worries me that the workers are still running in a different tokio runtime and it is necessitating extra machinery...

Reply:
After having a moment to think about this one, I realized the compiler worker shouldn't have any reason to exit anymore in normal execution. The new logic here should be a lot easier to wrap your head around.

Reply:
Yes - nice clean up! (But my comment about dual tokio runtimes remains. We need to fix that somehow. I hope to have time next week to work on it if you don't already get to it.) 

A tangental comment: the compiler worker (I think!) only issues FetchModuleMetaData ops. It would be nice to restrict it to those explicitly here:
https://github.com/denoland/deno/blob/744e56cb58f90acde7baf8811676021fb55b2f2a/cli/ops.rs#L146-L152